pkg/controllers/provisioningv2/cluster/controller.go

```diff
@@ -300,6 +319,23 @@ func (h *handler) createNewCluster(cluster *v1.Cluster, status v1.ClusterStatus,
 		})
 	}
 
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil {
+		clusterCopy := cluster.Spec.ClusterAgentDeploymentCustomization.DeepCopy()
+		spec.ClusterAgentDeploymentCustomization = &v3.AgentDeploymentCustomization{
+			AppendTolerations:            clusterCopy.AppendTolerations,
+			OverrideAffinity:             clusterCopy.OverrideAffinity,
+			OverrideResourceRequirements: clusterCopy.OverrideResourceRequirements,
+		}
+	}
+	if cluster.Spec.FleetAgentDeploymentCustomization != nil {
+		fleetCopy := cluster.Spec.FleetAgentDeploymentCustomization.DeepCopy()
+		spec.FleetAgentDeploymentCustomization = &v3.AgentDeploymentCustomization{
+			AppendTolerations:            fleetCopy.AppendTolerations,
+			OverrideAffinity:             fleetCopy.OverrideAffinity,
+			OverrideResourceRequirements: fleetCopy.OverrideResourceRequirements,
+		}
+	}
+
```

Comment:
```suggestion
	if cluster.Spec.ClusterAgentDeploymentCustomization != nil {
		spec.ClusterAgentDeploymentCustomization = cluster.Spec.ClusterAgentDeploymentCustomization.DeepCopy()
	}
	if cluster.Spec.FleetAgentDeploymentCustomization != nil {
		spec.FleetAgentDeploymentCustomization = cluster.Spec.FleetAgentDeploymentCustomization.DeepCopy()
	}

```
Is there any particular reason you were copying them in that way that's just not immediately obvious? i.e. I'd expect the types to be the same. Also, FWIW you don't need to nil check, `DeepCopy()` will return nil if the callee is nil:

ex.
```
func (in *Namespace) DeepCopy() *Namespace {
	if in == nil {
		return nil
	}
	out := new(Namespace)
	in.DeepCopyInto(out)
	return out
}
```

Reply:
I already discussed this with @jiaqiluo but doing that causes a compilation err because you are trying to assign a v3 struct to a v1 one. They are defined in different pkgs. That's why I assigned out the individual fields. I believe this answers/nullifies this comment as well https://github.com/rancher/rancher/pull/41203#discussion_r1180650651.

Reply:
Yup, thought of the same and noticed the types are different unfortunately.